# Importing clean dataframe: 3 binary columns and 39 numeric columns

In [1]:
import pandas as pd

In [2]:
d1 = pd.read_csv('./datasets/df_with_dummy.csv')

# Dummy and Numeric List

In [3]:
dummy_col = ['Foundation_PConc', 'Bsmt Qual_Ex', 'Kitchen Qual_Ex']

In [4]:
numeric_col = []
irrelevant_col = ['Unnamed: 0', 'SalePrice', 'Foundation_PConc', 'Bsmt Qual_Ex', 'Kitchen Qual_Ex']
for col in list(d1):
    if col not in irrelevant_col:
        numeric_col.append(col)
    
                                                                      
                                                                      

In [5]:
len(numeric_col)

39

# Polynomial for Train

In [6]:
from sklearn.preprocessing import StandardScaler, PolynomialFeatures

In [7]:
X = d1[numeric_col]
poly = PolynomialFeatures(include_bias= False)
X_train_poly = poly.fit_transform(X)

In [8]:
#creating a dataframe of our polynomial features
poly_df = pd.DataFrame(X_train_poly, columns = poly.get_feature_names(numeric_col))

## `poly_df` is the mother of polynomial features 

In [9]:
poly_df.head()

MS SubClass  Lot Frontage  Lot Area  Overall Qual  Overall Cond  \
0         60.0          68.0   13517.0           6.0           8.0   
1         60.0          43.0   11492.0           7.0           5.0   
2         20.0          68.0    7922.0           5.0           7.0   
3         60.0          73.0    9802.0           5.0           5.0   
4         50.0          82.0   14235.0           6.0           8.0   

   Year Built  Year Remod/Add  Mas Vnr Area  BsmtFin SF 1  BsmtFin SF 2  \
0      1976.0          2005.0         289.0         533.0           0.0   
1      1996.0          1997.0         132.0         637.0           0.0   
2      1953.0          2007.0           0.0         731.0           0.0   
3      2006.0          2007.0           0.0           0.0           0.0   
4      1900.0          1993.0           0.0           0.0           0.0   

      ...       Yr Sold^2  Yr Sold homeage  Yr Sold baths  Yr Sold garageage  \
0     ...       4040100.0          68340.0         5025.0            68340.0   
1     ...       4036081.0          26117.0         5022.5            24108.0   
2     ...       4040100.0         114570.0         2010.0           114570.0   
3     ...       4040100.0           8040.0         5025.0             6030.0   
4     ...       4040100.0         221100.0         4020.0           106530.0   

   homeage^2  homeage baths  homeage garageage  baths^2  baths garageage  \
0     1156.0           85.0             1156.0     6.25             85.0   
1      169.0           32.5              156.0     6.25             30.0   
2     3249.0           57.0             3249.0     1.00             57.0   
3       16.0           10.0               12.0     6.25              7.5   
4    12100.0          220.0             5830.0     4.00            106.0   

   garageage^2  
0       1156.0  
1        144.0  
2       3249.0  
3          9.0  
4       2809.0  

[5 rows x 819 columns]

In [10]:
corr_col = ['Overall Qual',
 'Year Built',
 'Year Remod/Add',
 'Mas Vnr Area',
 'Total Bsmt SF',
 '1st Flr SF',
 'Gr Liv Area',
 'Full Bath',
 'TotRms AbvGrd',
 'Garage Yr Blt',
 'Garage Cars',
 'Garage Area',
 'homeage',
 'baths',
 'garageage',
 'Lot Frontage Overall Qual',
 'Lot Frontage Mas Vnr Area',
 'Lot Frontage Gr Liv Area',
 'Lot Frontage Full Bath',
 'Lot Frontage Garage Cars',
 'Lot Frontage Garage Area',
 'Lot Frontage baths',
 'Lot Area Overall Qual',
 'Lot Area Garage Cars',
 'Overall Qual^2',
 'Overall Qual Overall Cond',
 'Overall Qual Year Built',
 'Overall Qual Year Remod/Add',
 'Overall Qual Mas Vnr Area',
 'Overall Qual BsmtFin SF 1',
 'Overall Qual Total Bsmt SF',
 'Overall Qual 1st Flr SF',
 'Overall Qual Gr Liv Area',
 'Overall Qual Full Bath',
 'Overall Qual Bedroom AbvGr',
 'Overall Qual Kitchen AbvGr',
 'Overall Qual TotRms AbvGrd',
 'Overall Qual Fireplaces',
 'Overall Qual Garage Yr Blt',
 'Overall Qual Garage Cars',
 'Overall Qual Garage Area',
 'Overall Qual Yr Sold',
 'Overall Qual baths',
 'Overall Cond Total Bsmt SF',
 'Overall Cond 1st Flr SF',
 'Overall Cond Gr Liv Area',
 'Overall Cond Garage Cars',
 'Overall Cond Garage Area',
 'Overall Cond baths',
 'Year Built^2',
 'Year Built Year Remod/Add',
 'Year Built Mas Vnr Area',
 'Year Built Total Bsmt SF',
 'Year Built 1st Flr SF',
 'Year Built Gr Liv Area',
 'Year Built Full Bath',
 'Year Built TotRms AbvGrd',
 'Year Built Garage Yr Blt',
 'Year Built Garage Cars',
 'Year Built Garage Area',
 'Year Built Yr Sold',
 'Year Built homeage',
 'Year Built baths',
 'Year Built garageage',
 'Year Remod/Add^2',
 'Year Remod/Add Mas Vnr Area',
 'Year Remod/Add Total Bsmt SF',
 'Year Remod/Add 1st Flr SF',
 'Year Remod/Add Gr Liv Area',
 'Year Remod/Add Full Bath',
 'Year Remod/Add TotRms AbvGrd',
 'Year Remod/Add Garage Yr Blt',
 'Year Remod/Add Garage Cars',
 'Year Remod/Add Garage Area',
 'Year Remod/Add Yr Sold',
 'Year Remod/Add homeage',
 'Year Remod/Add baths',
 'Year Remod/Add garageage',
 'Mas Vnr Area Total Bsmt SF',
 'Mas Vnr Area 1st Flr SF',
 'Mas Vnr Area Gr Liv Area',
 'Mas Vnr Area Full Bath',
 'Mas Vnr Area Bedroom AbvGr',
 'Mas Vnr Area TotRms AbvGrd',
 'Mas Vnr Area Fireplaces',
 'Mas Vnr Area Garage Yr Blt',
 'Mas Vnr Area Garage Cars',
 'Mas Vnr Area Garage Area',
 'Mas Vnr Area Yr Sold',
 'Mas Vnr Area baths',
 'BsmtFin SF 1 Full Bath',
 'BsmtFin SF 1 Garage Cars',
 'BsmtFin SF 1 baths',
 'Total Bsmt SF 1st Flr SF',
 'Total Bsmt SF Gr Liv Area',
 'Total Bsmt SF Full Bath',
 'Total Bsmt SF Bedroom AbvGr',
 'Total Bsmt SF TotRms AbvGrd',
 'Total Bsmt SF Fireplaces',
 'Total Bsmt SF Garage Yr Blt',
 'Total Bsmt SF Garage Cars',
 'Total Bsmt SF Garage Area',
 'Total Bsmt SF Yr Sold',
 'Total Bsmt SF baths',
 '1st Flr SF^2',
 '1st Flr SF Gr Liv Area',
 '1st Flr SF Full Bath',
 '1st Flr SF TotRms AbvGrd',
 '1st Flr SF Fireplaces',
 '1st Flr SF Garage Yr Blt',
 '1st Flr SF Garage Cars',
 '1st Flr SF Garage Area',
 '1st Flr SF Yr Sold',
 '1st Flr SF baths',
 'Gr Liv Area^2',
 'Gr Liv Area Full Bath',
 'Gr Liv Area TotRms AbvGrd',
 'Gr Liv Area Fireplaces',
 'Gr Liv Area Garage Yr Blt',
 'Gr Liv Area Garage Cars',
 'Gr Liv Area Garage Area',
 'Gr Liv Area Yr Sold',
 'Gr Liv Area baths',
 'Full Bath^2',
 'Full Bath TotRms AbvGrd',
 'Full Bath Fireplaces',
 'Full Bath Garage Yr Blt',
 'Full Bath Garage Cars',
 'Full Bath Garage Area',
 'Full Bath Yr Sold',
 'Full Bath baths',
 'Bedroom AbvGr Garage Cars',
 'Bedroom AbvGr Garage Area',
 'Kitchen AbvGr Garage Cars',
 'Kitchen AbvGr Garage Area',
 'Kitchen AbvGr homeage',
 'Kitchen AbvGr garageage',
 'TotRms AbvGrd Fireplaces',
 'TotRms AbvGrd Garage Yr Blt',
 'TotRms AbvGrd Garage Cars',
 'TotRms AbvGrd Garage Area',
 'TotRms AbvGrd Yr Sold',
 'TotRms AbvGrd baths',
 'Fireplaces Garage Cars',
 'Fireplaces Garage Area',
 'Fireplaces baths',
 'Garage Yr Blt^2',
 'Garage Yr Blt Garage Cars',
 'Garage Yr Blt Garage Area',
 'Garage Yr Blt Yr Sold',
 'Garage Yr Blt homeage',
 'Garage Yr Blt baths',
 'Garage Yr Blt garageage',
 'Garage Cars^2',
 'Garage Cars Garage Area',
 'Garage Cars Yr Sold',
 'Garage Cars baths',
 'Garage Area^2',
 'Garage Area Yr Sold',
 'Garage Area baths',
 'Yr Sold homeage',
 'Yr Sold baths',
 'Yr Sold garageage',
 'baths^2']
#list is from another notebook- but these are the columns that have a correlation higher than .50

In [11]:
len(poly_df[corr_col].columns)

164

# Train Scale Time 

In [12]:
#I want to standardize my new polynomial features 
X = poly_df[corr_col]

In [13]:
ss= StandardScaler()
ss.fit(X)  #learning the means and std for each column 
poly_scaled = ss.transform(X)

In [14]:
train_poly = pd.DataFrame(poly_scaled, columns = corr_col)

In [15]:
train_poly.shape

(2051, 164)

### When we transform test, we must use  poly_test_scaled = ss.transform(X) then test_poly = pd.DataFrame(poly_test_scaled, columns = corr_col)

In [16]:
d1[dummy_col].shape

(2051, 3)

In [17]:
train_sc_dum = train_poly.join(d1[dummy_col])

In [18]:
train_sc_dum.shape

(2051, 167)

In [19]:
train_sc_dum.head()

Overall Qual  Year Built  Year Remod/Add  Mas Vnr Area  Total Bsmt SF  \
0     -0.078644    0.142227        0.989479      1.092329      -0.741232   
1      0.622656    0.805126        0.609090      0.191491      -0.322705   
2     -0.779944   -0.620106        1.084576     -0.565901      -0.002130   
3     -0.779944    1.136575        1.084576     -0.565901      -1.500370   
4     -0.078644   -2.376787        0.418896     -0.565901      -0.850317   

   1st Flr SF  Gr Liv Area  Full Bath  TotRms AbvGrd  Garage Yr Blt  \
0   -1.108838    -0.040634   0.769779      -0.279441      -0.112447   
1   -0.634510     1.244529   0.769779       1.002738       0.737090   
2   -0.271195    -0.884084  -1.051232      -0.920531      -1.042893   
3   -1.060900    -0.110588   0.769779       0.361648       1.141632   
4   -0.841397    -0.108589   0.769779      -0.279441      -0.881076   

        ...         Garage Area^2  Garage Area Yr Sold  Garage Area baths  \
0       ...             -0.199323             0.008581           0.496281   
1       ...              0.182637             0.396922           0.848706   
2       ...             -0.925409            -1.053629          -1.083758   
3       ...             -0.487916            -0.339305           0.181616   
4       ...             -0.161367             0.050327           0.127913   

   Yr Sold homeage  Yr Sold baths  Yr Sold garageage   baths^2  \
0        -0.067210       1.160521           0.203566  1.138397   
1        -0.763445       1.158568          -0.686130  1.138397   
2         0.695100      -1.194594           1.133449 -1.046041   
3        -1.061526       1.160521          -1.049756  1.138397   
4         2.451726       0.375483           0.971731  0.202209   

   Foundation_PConc  Bsmt Qual_Ex  Kitchen Qual_Ex  
0                 0             0                0  
1                 1             0                0  
2                 0             0                0  
3                 1             0                0  
4                 1             0                0  

[5 rows x 167 columns]

In [20]:
train_sc_dum.to_csv('./datasets/train_features_for_gridsearch')

# Model 

In [21]:
from sklearn import metrics
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression, LassoCV, RidgeCV, ElasticNetCV
from sklearn.preprocessing import StandardScaler, PowerTransformer
from sklearn.metrics import r2_score

In [22]:
import numpy as np

In [23]:
#creating our x and y
X = train_sc_dum
y = d1['SalePrice']

In [24]:
#train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [25]:
#instantiate the models
lr = LinearRegression()
lasso = LassoCV()
ridge = RidgeCV()
elastic = ElasticNetCV()

# Cross Val Scores

In [26]:
cross_val_score(lr, X_train, y_train, cv= 10).mean()

0.8628502482458457

In [27]:
cross_val_score(ridge, X_train, y_train, cv= 10).mean()

0.8666884410559597

In [28]:
cross_val_score(elastic, X_train, y_train, cv= 10).mean()

/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:2053: FutureWarning: You should specify a va

0.6003718329968822

In [29]:
cross_val_score(lasso, X_train, y_train, cv= 10).mean()

/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:2053: FutureWarning: You should specify 

/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/model_selection/_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:492: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/Users/erinhwang/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/coordinate_descent.py:492: ConvergenceWarning: Objectiv

0.7825381459907106

# Proceed with Ridge

In [30]:
ridge.fit(X_train, y_train)

RidgeCV(alphas=array([ 0.1,  1. , 10. ]), cv=None, fit_intercept=True,
    gcv_mode=None, normalize=False, scoring=None, store_cv_values=False)

In [31]:
#cv scores
cross_val_score(ridge, X_train, y_train, cv=10).mean()

0.8666884410559597

In [32]:
#train score
ridge.score(X_train, y_train)

0.9283344627138861

In [33]:
#test score
ridge.score(X_test, y_test)

0.9031217146268262

In [34]:
from sklearn.metrics import mean_squared_error
import math

In [35]:
#train prediction 
train_pred = ridge.predict(X_train)

In [36]:
#mse train
mean_squared_error(y_train, train_pred)

459117493.74163485

In [37]:
#import numpy as np

In [38]:
#rmse train
np.sqrt(mean_squared_error(y_train, train_pred))

21427.027179280725

In [39]:
#test prediction 
test_pred = ridge.predict(X_test)

In [40]:
#mse test
mean_squared_error(y_test, test_pred)

579082877.2663937

In [41]:
#rmse test
np.sqrt(mean_squared_error(y_test, test_pred))

24064.140900235638

# Import Clean Test.csv data

In [42]:
test = pd.read_csv('./datasets/clean_test.csv')

In [43]:
test.head()

Unnamed: 0    Id        PID  MS SubClass  Lot Frontage  Lot Area  \
0           0  2658  902301120          190          69.0      9142   
1           1  2718  905108090           90          68.0      9662   
2           2  2414  528218130           60          58.0     17104   
3           3  1989  902207150           30          60.0      8520   
4           4   625  535105100           20          68.0      9500   

   Overall Qual  Overall Cond  Year Built  Year Remod/Add      ...        \
0             6             8        1910            1950      ...         
1             5             4        1977            1977      ...         
2             7             5        2006            2006      ...         
3             5             6        1923            2006      ...         
4             6             5        1963            1963      ...         

   Sale Type_COD  Sale Type_CWD  Sale Type_Con  Sale Type_ConLD  \
0              0              0              0                0   
1              0              0              0                0   
2              0              0              0                0   
3              0              0              0                0   
4              0              0              0                0   

   Sale Type_ConLI  Sale Type_ConLw  Sale Type_New  Sale Type_Oth  \
0                0                0              0              0   
1                0                0              0              0   
2                0                0              1              0   
3                0                0              0              0   
4                0                0              0              0   

   Sale Type_VWD  Sale Type_WD   
0              0              1  
1              0              1  
2              0              0  
3              0              1  
4              0              1  

[5 rows x 278 columns]

In [44]:
test[numeric_col].shape

(879, 39)

In [45]:
test[dummy_col].shape

(879, 3)

# Polynomial Transformation of Test

In [46]:
X = test[numeric_col]
poly = PolynomialFeatures(include_bias= False)
X_test_poly = poly.fit_transform(X)

In [47]:
#creating a dataframe of our polynomial features
poly_df = pd.DataFrame(X_test_poly, columns = poly.get_feature_names(numeric_col))

In [48]:
poly_df.shape

(879, 819)

In [49]:
#using same list from model 
poly_df[corr_col].shape

(879, 164)

# Test Scale Time

In [50]:
#I want to standardize my new polynomial features 
X = poly_df[corr_col]

In [51]:
poly_test_scaled = ss.transform(X)

In [52]:
test_poly = pd.DataFrame(poly_test_scaled, columns = corr_col)

In [53]:
test_poly.shape

(879, 164)

# Lets append the dummies

In [54]:
test[dummy_col].shape

(879, 3)

In [55]:
test_features = test_poly.join(test[dummy_col])

In [56]:
test_features.shape

(879, 167)

# Predict using Ridge

In [57]:
pred = ridge.predict(test_features)

# Before we transfer our predictions, lets make sure our model is right
- In the `Ames_GridSearch1` notebook, we will grid search using elastic net to see best fit and determine if we should consider the Lasso and or Ridge model. 

*Note: Please look at the `Ames_GridSearch1` notebook to see how we executed our GrideSearch model*

# Model Conclusion

By executing the `cross_val_score`, proceeding with the `Ridge` was the best model for us. Here are the following regression metrics for Ridge:

|Metric|Score|
|---|---|
|Train Score|.9283|
|Test Score|.9031|
|RMSE Train|21,427|
|RMSE Test|24,064|

By executing `GridSearchCV`, we used `ElasticNet` as the model since it involves both Lasso and Ridge. GridSearch will tells us the best ratio between Lasso and Ridge. After looking at the recommended parameters, we computed regression metrics. Here are the following metrics for elastic. 

|Metric|Score|
|---|---|
|Train Score|.9280|
|Test Score|.8978|
|RMSE Train|21,467|
|RMSE Test|24,711|

**Final:** For the features we used in our regression model, it is best to proceed with the ridge model since the overall test score and RMSE score have a better overall fit with the true sale price of homes. 

# Create CSV to import results

In [58]:
test['Id'].shape

(879,)

In [59]:
pred.shape

(879,)

In [60]:
submission = pd.DataFrame({
    'Id': test['Id'],
    'SalePrice': pred
})

submission.to_csv('./datasets/submissionv_thattooklong4.csv', index= False)